### PR TITLE
Switch to multimap based nfd_map due to compile time issues

### PR DESCRIPTION
**Pull request by iamlemec**
Fixes issues with #5740. Yields same tokenizer outcomes as before but brings compile time back to normal. Performance also seems to be unchanged. Though in testing I did notice that #5740 itself does appear to reduce perfomance substantially, so there is hopefully a better long-term solution here.

### CODE REVIEW

File: unicode.h [hunk 1]
```diff
@@ -223,266 +224,409 @@ static const std::vector<std::pair<uint32_t, uint32_t>> control_ranges = {
...
-{0x2F9EF, {0x4995}}, {0x2F9F0, {0x95B7}}, {0x2F9F1, {0x28D77}}, {0x2F9F2, {0x49E6}}, {0x2F9F3, {0x96C3}}, {0x2F9F4, {0x5DB2}}, {0x2F9F5, {0x9723}}, {0x2F9F6, {0x29145}}, {0x2F9F7, {0x2921A}},
-{0x2F9F8, {0x4A6E}}, {0x2F9F9, {0x4A76}}, {0x2F9FA, {0x97E0}}, {0x2F9FB, {0x2940A}}, {0x2F9FC, {0x4AB2}}, {0x2F9FD, {0x29496}}, {0x2F9FE, {0x980B}}, {0x2F9FF, {0x980B}}, {0x2FA00, {0x9829}},
-{0x2FA01, {0x295B6}}, {0x2FA02, {0x98E2}}, {0x2FA03, {0x4B33}}, {0x2FA04, {0x9929}}, {0x2FA05, {0x99A7}}, {0x2FA06, {0x99C2}}, {0x2FA07, {0x99FE}}, {0x2FA08, {0x4BCE}}, {0x2FA09, {0x29B30}},
-{0x2FA0A, {0x9B12}}, {0x2FA0B, {0x9C40}}, {0x2FA0C, {0x9CFD}}, {0x2FA0D, {0x4CCE}}, {0x2FA0E, {0x4CED}}, {0x2FA0F, {0x9D67}}, {0x2FA10, {0x2A0CE}}, {0x2FA11, {0x4CF8}}, {0x2FA12, {0x2A105}},
-{0x2FA13, {0x2A20E}}, {0x2FA14, {0x2A291}}, {0x2FA15, {0x9EBB}}, {0x2FA16, {0x4D56}}, {0x2FA17, {0x9EF9}}, {0x2FA18, {0x9EFE}}, {0x2FA19, {0x9F05}}, {0x2FA1A, {0x9F0F}}, {0x2FA1B, {0x9F16}},
-{0x2FA1D, {0x2A600}},
+typedef std::pair<uint32_t, uint32_t> CodePos;
```
Comment (cebtenzzre) [inline unicode.h:227]: You don't need CodePos - std::multimap will work just as well if the keys are strictly (bitwise) equal.

**Comment by cebtenzzre** [comment]
Compile times would likely be reduced by making unicode.h a proper .cpp file instead of a header full of static tables and functions - especially if you include the time to compile the tests.

The rust implementation that apage43 linked uses minimally perfect hash tables for faster lookup (all keys are guaranteed to have a different hash so collisions are impossible) - if lookup is a bottleneck, maybe it would be worth trying to implement something similar.

File: unicode.h (hunk 1 above)
Comment (iamlemec) [reply unicode.h]: Oh wow, yeah that's much simpler!

**Comment by iamlemec** [comment]
Turns out we were spending an inordinate amount of time creating `std::locale("utf-8")` with every single character in `to_lower`. Made it static and we're pretty close to the old performance now.

**Review by cebtenzzre — APPROVED** [review]

**Review by ggerganov — APPROVED** [review]
It's still not back to original level, but much better:

```bash
# before
real	0m5.506s
user	0m5.352s
sys	    0m0.101s

# after
real	0m6.396s
user	0m6.216s
sys	0m0.109s
```

**Pull request merged**